add parameterized types

ygdrasil-io opened the pull request.

ygdrasil-io committed:
Add support for parameterized WebIDL types

Extended the parser and type model to handle parameterized types such as `sequence`, `record`, `FrozenArray`, and `Promise`. Updated validation logic to recognize these types and parse their parameters correctly. Adjusted the `IdlType` class to store parameterized type information.

ygdrasil-io committed:
Refactor whileWithMaxTry function for clarity.

Renamed parameters and improved variable naming to enhance readability. Adjusted error messaging to align with updated parameter names for better context.

ygdrasil-io committed:
Filter out unsupported parameterized types in imports.

Previously, all imports were processed without regard for parameterized types, potentially causing issues. This update adds a filter to exclude unsupported parameterized types, improving stability and error handling. A TODO comment is included for future support of these types.

ygdrasil-io committed:
Update test to reflect increase in attribute count

Previously, the tested interface contained 2 attributes, but this has been updated to 7. The change aligns the test with the updated interface definition.

ygdrasil-io commented:
waiting https://github.com/fabmax/webidl-util/pull/4 to be merged

ygdrasil-io committed:
"Fix indentation in JniJavaGenerator.kt"

Corrected inconsistent indentation in the lambda to improve code readability. This change does not alter functionality but ensures the code adheres to proper formatting standards.

ygdrasil-io committed:
Fix redundant condition handling in type parsing logic

Removed unnecessary conditional block checking for parameterized types within angled brackets. Simplifies code by consolidating checks and reducing duplication, improving maintainability.

ygdrasil-io committed:
Remove unused `endWith` function from WebIdlStream.

fabmax commented on webidl-util/src/main/kotlin/de/fabmax/webidl/parser/WebIdlStream.kt:
```diff
@@ -128,14 +123,42 @@ class WebIdlStream {
+            }
+
+            val typeParams = typeName.substring(typeName.indexOf("<") + 1, typeName.lastIndexOf(">"))
+            typeName = typeName.substring(0, typeName.indexOf("<"))
```
Personally, I prefer using the `substringBefore` / `substringAfter` functions for these kind of things. E.g. something like this:
```
    val typeParams = typeName.substringAfter('<').substringBeforeLast('>')
    typeName = typeName.substringBefore('<')
```
But that's only a personal preference so feel free to keep it this way 😄

ygdrasil-io replied:
I agree, that part lack also of Unit Test on the result, working on it.

ygdrasil-io committed:
Simplify string handling in type parsing logic

Refactor type parameter extraction to use more concise Kotlin functions `substringAfter` and `substringBeforeLast`.

ygdrasil-io committed:
Refactor `IdlType` to support multiple parameter types.

Updated `IdlType.parameterType` from a single string to a list to handle multiple type parameters. Adjusted parsing logic in `WebIdlStream` and updated tests to validate the new structure. Ensures compatibility with complex type definitions like `record` and `FrozenArray`.

ygdrasil-io committed:
Fix typo in IdlType field and update corresponding test.

Renamed the `parameterType` field to `parameterTypes` in the `IdlType` class for consistency and clarity. Updated relevant test assertions to reflect this change.
